baselines/fedpara/fedpara/models.py

```diff
@@ -0,0 +1,391 @@
+        for buffer in self.buffers():
+            buffer_size += buffer.nelement() * buffer.element_size()
+        size_all_mb = (param_size + buffer_size) / 1024**2
+        return total_trainable_params, size_all_mb
+
+    def forward(self, input):
+        """Forward pass."""
+        x = self.features(input)
+        x = x.view(x.size(0), -1)
+        x = self.classifier(x)
+        return x
+
+
+# Create an instance of the VGG16GN model with Group Normalization,
+# custom Conv2d, and modified classifier
+def test(
+    net: nn.Module, test_loader: DataLoader, device: torch.device
+) -> Tuple[float, float]:
+    """Evaluate the network on the entire test set.
+
+    Parameters
+    ----------
+    net : nn.Module
+        The neural network to test.
+    test_loader : DataLoader
+        The DataLoader containing the data to test the network on.
+    device : torch.device
+        The device on which the model should be tested, either 'cpu' or 'cuda'.
+
+    Returns
+    -------
+    Tuple[float, float]
+        The loss and the accuracy of the input model on the given data.
+    """
+    if len(test_loader.dataset) == 0:
+        raise ValueError("Testloader can't be 0, exiting...")
+
+    criterion = torch.nn.CrossEntropyLoss()
+    correct, total, loss = 0, 0, 0.0
+    net.eval()
+    with torch.no_grad():
+        for images, labels in tqdm(test_loader, "Testing ..."):
+            images, labels = images.to(device), labels.to(device)
+            outputs = net(images)
+            loss += criterion(outputs, labels).item()
+            _, predicted = torch.max(outputs.data, 1)
+            total += labels.size(0)
+            correct += (predicted == labels).sum().item()
+    loss /= len(test_loader.dataset)
+    accuracy = correct / total
+    return loss, accuracy
+
+
+def train(  # pylint: disable=too-many-arguments
+    net: nn.Module,
+    trainloader: DataLoader,
+    device: torch.device,
+    epochs: int,
+    hyperparams: Dict[str, Scalar],
+    round: int,
+) -> None:
+    """Train the network on the training set.
+
+    Parameters
+    ----------
+    net : nn.Module
+        The neural network to train.
+    trainloader : DataLoader
+        The DataLoader containing the data to train the network on.
+    device : torch.device
+        The device on which the model should be trained, either 'cpu' or 'cuda'.
+    epochs : int
+        The number of epochs the model should be trained for.
+    hyperparams : Dict[str, Scalar]
+        The hyperparameters to use for training.
+    """
+    lr = float(hyperparams["eta_l"]) * float(hyperparams["learning_decay"]) ** (
+        round - 1
+    )
+    print(f"Learning rate: {lr}")
+    criterion = torch.nn.CrossEntropyLoss()
+    optimizer = torch.optim.SGD(
+        net.parameters(),
+        lr=lr,
+        momentum=hyperparams["momentum"],
+        weight_decay=hyperparams["weight_decay"],
+    )
+    net.train()
+    for _ in tqdm(range(epochs), desc="Local Training ..."):
```

Comment:
Maybe it's worth removing `tqdm` for the clients? (It progress bars don't looks so nice when multiple processes update their own...). Another option is to have a flag in the configs to toggle it on/off (but better keep it off by default)

Reply:
done!